### PR TITLE
TRITON-2184 CloudAPI should allow exec commands on zones.

### PR DESCRIPTION
**Pull request by kusor**
Co-authored-by: Alex Wilson <alex@uq.edu.au>

### CODE REVIEW

File: lib/endpoints/exec.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        res.send(400, new restify.RestError({
+            statusCode: 400,
+            restCode: 'MachineStopped',
+            message: 'Only running machines can execute commands'
+        }));
+        next();
+        return;
+    }
+
+    if (!req.params.argv || !Array.isArray(req.params.argv)) {
```
Comment (twhiteman) [inline lib/endpoints/exec.js:42]:
```suggestion
    if (!Array.isArray(req.params.argv)) {
```

File: test/common.js [hunk 2]
```diff
@@ -72,7 +72,7 @@ var BHYVE_128_FLEX_PACKAGE = Object.assign({}, BHYVE_128_PACKAGE, {
     uuid: '8b3b6796-e5b1-47ce-bb49-ac9dcba7c6eb',
     brand: 'bhyve',
     name: 'sdc_128_flex_bhyve',
-    quota: 50 * 1024, // 50 GiB
+    quota: 20 * 1024, // 50 GiB
```
Comment (twhiteman) [inline test/common.js:75]: Why is this changing? I'm not 100% sure, but I think some depends on this being 50Gib.

**Review by twhiteman — CHANGES_REQUESTED** [review]
Mostly there, just needs some tweaks:
* bump package.json
* add /exec documentation
* add a /exec machine test

**Comment by arekinath** [comment]
In case you guys care about this, we stopped using this in-line exec endpoint in favour of a websocket, since dealing with timeouts on long-running commands through this interface is pretty painful. https://github.com/arekinath/sdc-cloudapi/commit/8ee0fa4ce9d16ee3b1b50ef7f6010cbd28fe14b7 is the current revision of our patch with the websocket bit (it's very similar to the VNC websocket endpoint, though probably needs some of the same cleanup which the VNC one had done to it when it was upstreamed).

**Comment by kusor** [comment]
> In case you guys care about this, we stopped using this in-line exec endpoint in favour of a websocket, since dealing with timeouts on long-running commands through this interface is pretty painful. [arekinath@8ee0fa4](https://github.com/arekinath/sdc-cloudapi/commit/8ee0fa4ce9d16ee3b1b50ef7f6010cbd28fe14b7) is the current revision of our patch with the websocket bit (it's very similar to the VNC websocket endpoint, though probably needs some of the same cleanup which the VNC one had done to it when it was upstreamed).

I definitely like the idea of using a websocket, yeah! I'm gonna make a new issue to update the end-point to use websockets.

Btw, now that I've got your attention: any special preference to use a FSM? I'd personally go ahead with something simpler for this kind of thing, though.

**Comment by Smithx10** [comment]
@arekinath  Curious if you brought over other docker api features into cloudapi zones like the /archive endpoint for uploading tar.gz's into zones?  Might be worth adding this functionality.

https://github.com/joyent/sdc-docker/blob/master/lib/endpoints/containers.js#L838

**Comment by arekinath** [comment]
@kusor I use the FSMs because I don't trust myself to cover all the async error cases and racey bits without them. If you're confident you can cover everything without them, go for it.

**Comment by arekinath** [comment]
@Smithx10 Nah, we haven't bothered with that. We just use exec to run `tar | curl` or `curl | tar` if we need to move things in and out with this mechanism at the moment. It seemed easier to do in parallel across a lot of machines (just sign a Manta URL for all of them to upload to and hand them out in the `curl` commands, instead of dealing with a whole bunch of high bandwidth tar streams all coming through cloudapi and the client that requested the tarballs at once).

File: test/common.js (hunk 2 above)
Comment (bahamat) [reply test/common.js:75]: I'd bias for keeping this value as-is, but if there's sufficient rationale for reducing the quota I'm not inherently opposed. But in any case the comment should match the value.

**Pull request closed without merging**